app/Helpers/database/leaderboard.php

```diff
@@ -2,134 +2,165 @@
+    $retVal['RankInfo'] = GetLeaderboardRankingJSON($user->User, $lbID, (bool) $leaderboard->LowerIsBetter);
+
     return $retVal;
 }
 
-function removeLeaderboardEntry(string $user, int $lbID, ?int &$score): bool
-{
-    sanitize_sql_inputs($user);
-
-    $query = "SELECT le.Score, ld.Format FROM LeaderboardEntry AS le
-              LEFT JOIN LeaderboardDef AS ld ON ld.ID = le.LeaderboardID
-              LEFT JOIN UserAccounts AS ua ON ua.ID = le.UserID
-              WHERE ua.User = '$user' AND ld.ID = $lbID ";
-
-    $dbResult = s_mysql_query($query);
-    $data = mysqli_fetch_assoc($dbResult);
-
-    if (!$data) {
-        return false;
+/**
+ * @deprecated
+ * This is intended only for double writes.
+ * Delete when LeaderboardEntry is dropped in favor of leaderboard_entries.
+ */
+function writeLegacyLeaderboardEntry(
+    User $user,
+    Leaderboard $leaderboard,
+    int $newEntry,
+): void {
+    $existingLegacyLeaderboardEntry = LeaderboardEntryLegacy::where('LeaderboardID', $leaderboard->id)
+        ->where('UserID', $user->id)
+        ->first();
+
+    if ($existingLegacyLeaderboardEntry) {
+        // If the user is submitting a better score than their current entry,
+        // we'll override the current entry with their new score.
+        $comparisonOp = $leaderboard->LowerIsBetter === 1 ? '<' : '>';
+        $hasBetterScore =
+            ($comparisonOp === '<' && $newEntry < $existingLegacyLeaderboardEntry->Score)
+            || ($comparisonOp === '>' && $newEntry > $existingLegacyLeaderboardEntry->Score)
+        ;
+
+        if ($hasBetterScore) {
+            // Update the player's entry.
+            // This has to use `update()` because the table has a composite primary key.
+            LeaderboardEntryLegacy::where('LeaderboardID', $leaderboard->id)
+                ->where('UserID', $user->id)
+                ->update(['Score' => $newEntry, 'DateSubmitted' => now()]);
+        }
+    } else {
+        // No existing leaderboard entry. Let's insert a new one.
+        LeaderboardEntryLegacy::create([
+            'LeaderboardID' => $leaderboard->id,
+            'UserID' => $user->id,
+            'Score' => $newEntry,
+            'DateSubmitted' => now(),
+        ]);
     }
+}
 
-    $score = ValueFormat::format((int) $data['Score'], $data['Format']);
+function removeLeaderboardEntry(User $user, int $lbID, ?int &$score): bool
+{
+    $leaderboardEntry = LeaderboardEntry::with('leaderboard')
+        ->where('leaderboard_id', $lbID)
+        ->where('user_id', $user->id)
+        ->first();
 
-    $userID = getUserIDFromUser($user);
-    if ($userID === 0) {
+    if (!$leaderboardEntry) {
         return false;
     }
 
-    $query = "DELETE FROM LeaderboardEntry
-              WHERE ( LeaderboardID = $lbID AND UserID = $userID )";
+    $score = ValueFormat::format($leaderboardEntry->score, $leaderboardEntry->leaderboard->Format);
```

Comment:
I realize this is existing code, but `ValueFormat::format` returns a string. `$score` is an out `int`. The function signature should be updated.

Reply:
Good catch, I naively assumed this would be automatically caught and didn't think to check the function signature. Updated in latest.